[style] Add RTL support

alitaheri opened the pull request:
This is only a concept. I've added basic support plus some fixes.

Surely, I could implement the features better, but I've been away for a while 😅 So I'm scratching my head. Share your ideas guys 😁 

**What it does:**

1. build a new `jss` with the `jss-rtl` plugin included.
2. patch `withStyle` so can it can opt-in and out based on `theme.direction`
3. Add visual test (like dark/light theme) to docs
4. Patch some doc components so that only components get affected
5. `TouchRipple` shouldn't be affected so I opt that one out
6. `transform-origin` should be `top right` when `rtl` so I patched that for `InputLabel`

**What remains:**

1.  Components

- [x] Drawers
- [ ] ~~Buttons in Drawers~~ Inconclusive, Possibly a chrome bug.
- [x] Stepper
- [x] Snackbars 
- [x] Table Pagination
- [x] Tabs ~~(Almost done, indicator remaining)~~
- [x] Menu Positioning
- [x] Regression tests

![capture d ecran 2017-10-07 a 21 28 21](https://user-images.githubusercontent.com/3165635/31311157-891249f8-aba6-11e7-8bc0-c1d7b4e538f6.png)

Closes #6874

oliviertassinari commented on docs/src/modules/components/AppWrapper.js:
```diff
@@ -48,14 +48,18 @@ class AppWrapper extends React.Component<any, any> {
       }
+
+      if (nextProps.rtl !== this.props.rtl) {
+        document.body.dir = nextProps.rtl ? 'rtl' : 'ltr';
```
👍

oliviertassinari commented on docs/src/modules/components/AppFrame.js:
```diff
@@ -199,6 +205,15 @@ class AppFrame extends React.Component<any, any> {
+                aria-label="change direction"
+                onClick={this.handleToggleDirection}
+              >
+                {rtl ? <FormatTextdirectionLToR /> : <FormatTextdirectionRToL />}
```
In the long run, I want to add a dynamic palette section too so people can try out what it looks like with different primary, secondary colors. It's fine this way but it won't scale. We gonna need to find a different solution for the dynamic palette color.

alitaheri replied:
@oliviertassinari It's not very hard to implement I've done something similar before. We can keep a raw theme object (one you can pass to createMuiTheme) with an action `SET_THEME` any deep change will cause the recreation of the entire object. lodash has useful utils `get` and `set` for this.

I'd say we make an issue for that. It's not in the same scope as this change.

oliviertassinari replied:
Oh no no, I think there is a misunderstanding. We already have an issue for this. Also, we don't need any lodash function as it's a matter of call the `getMuiTheme` with the right arguments. My point was mostly about the docs UI/UX. You can ignore my comment for this PR.

NaamanTsur commented:
I've noticed that in the selected menus and max height menus examples, the menus are not in the correct position in RTL.

alitaheri commented:
@NaamanTsur Thanks 👍 👍 Please report any other anomalies you see.

NaamanTsur commented:
Alright I found a few more:

- Circular progress not going the other direction (not sure if it's suppose to)
- The multiple select demo has the same problem as the menu demos (might be because of the MenuItem component)
- Snackbars in the positioned demo are appearing in the mirrored position
- Native pickers are positioned LTR (don't know if they can or suppose to be in RTL)

smmoosavi commented:
I found some jss created only with preset (e.g. [getContext.js](https://github.com/alitaheri/material-ui/blob/e9e104eb7e9f0668259a792c45a5950878ea8385/docs/src/modules/styles/getContext.js#L27-L28))

alitaheri commented:
@NaamanTsur 

1. It's going clockwise, our clock goes clockwise too 😅 . Let's leave at that 
2. Yeah, I took note 👍 
3. 👍 👍 
4. They don't and they don't matter. They aren't very good for rtl anyway. It's best not used in an rtl environment.

@smmoosavi 

That doesn't matter, it's for server-side. rtl is only used on the client. better not put a plugin where it's not needed.

oliviertassinari commented:
> That doesn't matter, it's for server-side. rtl is only used on the client. better not put a plugin where it's not needed.

@alitaheri Server-side or not, if it's used on the client side, we need it for the server side too.

alitaheri commented:
@oliviertassinari Ok, I'll add it, but I'll have to disable it so it doesn't transform the styles. Otherwise we'll get mismatching hashes all the time. ( I think it's best be there too, I forgot about the `flip` prop 😅 )

P.S. You removed the demo for the mini drawer in your last commit, I added that back

oliviertassinari commented:
@alitaheri Good catch! I have added it back. I want to push before the end of the day.

alitaheri commented:
@oliviertassinari Since `scrollLeft` is really [messed up](https://github.com/othree/jquery.rtl-scroll-type#3-types-of-scrollleft-scrollwidth--100) I had to solve it in a very ugly way! I'll refactor the added code into another repository to clean up the code. But have a look at my solution.

@NaamanTsur  @smmoosavi Please checkout my work so far and see if you can catch any issues, thanks 😁 

I'm almost done

NaamanTsur commented:
Great work on the implementation so far 😃.
More stuff I found:
1. In the cards examples, the UI controls demo has it's previous and next button icons flipped the wrong way.
2. There's the weird button flicker in the drawer but i assume it's what you mentioned in your list. Oddly enough only the temporary drawer and the responsive drawer demo works fine. I guess if you can't solve it at the moment an issue for it should be opened.
3. The popovers's transformOrigin seems to be flipped, though it might be just me not knowing how it works so i may be wrong.
4. The query demo in the progress examples is going the wrong way.
5. In the tooltip examples the positioned tooltips example is technically correct though adding conditions to have the button text indicate the correct position (left being right) would be nice.
6. If you thought the last issue was nit-picky, in the simple tooltips demo the text shown is lying in RTL 😈.

Well that's all I managed to find! Again good job so far 😄

alitaheri commented:
@NaamanTsur 

1. I am aware of that, it's docs issue. I guess it doesn't hurt to fix it only to demonstrate that one should be careful around icons when dealing with rtl.
2. yup I'm aware of that 😑 😑 Can't seem to find the cause, I'm guessing it has to do with the scrollbar
3. I'll have another look at it.
4. It's supposed to go the other way 😅 Check the ltr version.
5. Yeah I thought of that, I'm not sure. I'll think about about it 👍 👍 

About the opting out, I saw it broke too 😱 😱 I guess it's a rebase regression. I'll have to fix it. But I think your idea is better 👍 👍 I should use a card

Thanks a lot @NaamanTsur, it was a very useful review 👍 👍

alitaheri commented:
@NaamanTsur 

1. Fixed
2. Later, No idea what is causing it!
3. It's working as expected
4. ...
5. Start and End are direction variant so, tooltip should flip 👍 👍 
6. I fixed it in a way that it won't be a lie any more 😁 

@oliviertassinari I'm done. have a final look at this please. Be thorough. I'm sure a lot of things can be improved.

alitaheri commented:
@oliviertassinari I guess extra 1KB is acceptable for a big feature such as `Right-To-Left`, right? 😅 😅

NaamanTsur commented:
I've noticed that whenever you flip from LTR to RTL and from RTL to LTR in the home screen the opening of the drawer is coming from the wrong direction but only once.

oliviertassinari commented:
> I guess extra 1KB is acceptable for a big feature

It's fine! Let me know when this is ready. I will have a deeper look at the PR :).

alitaheri commented:
@NaamanTsur Not very easy to fix, it's transition related. To fix it, I'll have to look out for a change in direction and disable the transition for a bit. This is not a good idea. makes the code very complicated for a use-case that probably never happens (you don't switch from rtl to ltr on a regular basis).

@oliviertassinari I'm done here. you can start reviewing, also please ping others to join in. Not very sure who the current collaborators are, 😅 😅 😅

oliviertassinari commented:
> I'm done here. you can start reviewing, also please ping others to join in. Not very sure who the current collaborators are

@alitaheri Things have changed a bit since you contributed here. I have taken the lead on the project, we have some new great collaborators :)
- @mbrookes As always, won't let any grammatical mistakes passed.
- @rosskevin is an awesome help moving the code base to flow and more addressing more core issues
- @sebald is an awesome help improving the TypeScript definitions, I know you love TypeScript :p.
- @kgregory Has been improving some of our components, as well as his coworker at @STORIS 
- @leMaik and his coworkers at @TeamWertarbyte have been implementing awesome components on top of Material-UI, he starts helping us here too :).
- @agamrafaeli Helped to reach a 99% test coverage.
- @nathanmarks Started this whole v1 effort!
- @kof has been helping us moving from our previous inline-style solution. We have made 3 iterations so far, might start the 4th one by relying more on react-jss.

And much more, 504 to be exact, thank you so much all!

I'm gonna prioritize the review of this PR, so we don't have too much rebase to handle.

alitaheri commented:
@oliviertassinari wow, nice team 😍 😍 You've all done a marvelous job on `v1` 👍 👍  

> I know you love TypeScript

Yeah I do 😊 

Don't worry about the rebases, Take your time. this feature adds complexity to the project. I'd rather make sure we minimize the impact.

sebald commented:
@alitaheri @oliviertassinari  Guess a view components got a `rtl` props :) Should we add them with this PR or later?

alitaheri commented:
@sebald Which component? I didn't find any `rtl` prop O.o

smmoosavi commented:
https://material-ui-1dab0.firebaseapp.com/api/table-cell/
> numeric: If true, content will align to the right.

numbers are always ltr. so maybe we need some logic to keep numeric table cells always align to the right

oliviertassinari commented:
Let's get this PR into the `v1-beta` branch!

alitaheri commented:
@oliviertassinari Please review the non-conflicting code until I fix them 😁

oliviertassinari commented:
@alitaheri Do you mind if I squash down, rebase and do a second commit with my changes?

alitaheri commented:
@oliviertassinari Not at all, go to town 👍 🚌 💥 😎 😎

alitaheri commented:
@oliviertassinari Nice changes 😅 I guess I've been away for a while. MIssed a lot of changes 😞

oliviertassinari commented:
@alitaheri I'm not done yet 😁 .

oliviertassinari commented:
@alitaheri Big thanks!!! Feel free to submit new PRs on top of this one if you find new issues. Let's tweet about it :).

smmoosavi commented:
we are waiting for new release :)
